File: rtui/src/data/github_sync.rs
//! Import existing GitHub review threads into the local Store so they render on the diff,
//! Conversation, and Comments views. Idempotent: comments already present (by github_id)
//! are refreshed; local replies (no github_id) are preserved.

use super::source::Source;
use super::store::{Comment, Store};

fn now() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    format!("{secs:039}-000000000000")
}

/// Import (or refresh) all of a PR's review threads into `store`. Returns count imported.
pub fn import(store: &mut Store, source: &Source) -> usize {
    if source.threads.is_empty() {
        return 0;
    }
    // Existing GH comments by their node id.
    let seen: std::collections::HashMap<String, String> = store
        .comments
        .values()
        .filter_map(|c| c.github_id.clone().map(|g| (g, c.id.clone())))
        .collect();
    let mut imported = 0;

    for t in &source.threads {
        let side = if t.side == "LEFT" { "LEFT" } else { "RIGHT" };
        let line = if t.line > 0 { t.line as u32 } else { 1 };
        let status = if t.resolved { "resolved" } else { "draft" };
        let mut root_id: Option<String> = None;

        for (i, cm) in t.comments.iter().enumerate() {
            if let Some(existing_id) = seen.get(&cm.id) {
                // refresh upstream-authoritative fields; keep local edits/replies intact
                if let Some(c) = store.comments.get_mut(existing_id) {
                    c.body = cm.body.clone();
                    if t.resolved {
                        c.status = "resolved".into();
                    }
                    c.outdated = t.outdated;
                    if i == 0 {
                        root_id = Some(c.id.clone());
                    }
                }
                continue;
            }
            let is_root = i == 0;
            let kind = if cm.body.contains("```suggestion") {
                "suggestion"
            } else {
                "normal"
            };
            let id = super::store::new_uuid();
            let c = Comment {
                id: id.clone(),
                file: t.path.clone(),
                side: side.into(),
                line_start: line,
                line_end: line,
                body: cm.body.clone(),
                origin: "github".into(),
                status: status.into(),
                kind: kind.into(),
                suggestion_text: None,
                in_reply_to: if is_root { None } else { root_id.clone() },
                github_id: Some(cm.id.clone()),
                gh_thread_id: if is_root { Some(t.id.clone()) } else { None },
                author: if cm.author.is_empty() {
                    "github".into()
                } else {
                    cm.author.clone()
                },
                created_at: if cm.created_at.is_empty() {
                    now()
                } else {
                    cm.created_at.clone()
                },
                hidden: false,
                reactions: Default::default(),
                anchor_text: None,
                outdated: t.outdated,
                workflow_state: if t.resolved { "resolved" } else { "unresolved" }.into(),
                assessment: String::new(),
                label: String::new(),
                priority: 0,
                action_owner: "author".into(),
                implementation_commit: None,
                implementation_worktree: None,
                push_state: String::new(),
                validation: vec![],
            };
            if is_root {
                root_id = Some(id.clone());
            }
            store.comments.insert(id, c);
            imported += 1;
        }
    }
    store.save();
    imported
}

File: lua/review/comments/github_sync.lua
-- review.nvim :: import GitHub review threads into the local store.
--
-- Idempotent: comments already present (matched by github_id) are refreshed from
-- upstream (github_id fields are upstream-authoritative, gap #8) without clobbering
-- local replies (which have no github_id).

local anchor = require("review.comments.anchor")
local util = require("review.util")

local M = {}

local function nullable(value)
  if value == vim.NIL then return nil end
  return value
end

--- Index existing comments by github_id.
---@param store table
---@return table<string, table>
local function by_github_id(store)
  local idx = {}
  for _, c in pairs(store.comments) do
    if c.github_id then
      idx[c.github_id] = c
    end
  end
  return idx
end

local function reaction_counts(groups)
  local out = {}
  for _, group in ipairs(groups or {}) do
    local count = group.users and tonumber(group.users.totalCount) or 0
    if group.content and count > 0 then out[group.content] = count end
  end
  return out
end

--- Import (or refresh) all GitHub review threads into `store`.
---@param source table
---@param store table
---@return integer imported
function M.import(source, store)
  local threads, fetch_err = source:threads()
  if fetch_err then return 0, fetch_err end
  if not threads or #threads == 0 then
    return 0
  end
  local meta = source:metadata()
  local existing = by_github_id(store)
  local now = os.date("!%Y-%m-%dT%H:%M:%SZ")
  local imported, refreshed = 0, 0

  for _, t in ipairs(threads) do
    local nodes = (t.comments and t.comments.nodes) or {}
    local root_local_id = nil
    for i, cm in ipairs(nodes) do
      local gid = nullable(cm.id)
      local side = (t.diffSide == "LEFT") and "LEFT" or "RIGHT"
      local line = tonumber(nullable(cm.line) or nullable(cm.originalLine)
        or nullable(t.line) or nullable(t.originalLine)) or 1
      local path = nullable(cm.path) or nullable(t.path)
      if gid and path then
        local prev = existing[gid]
        if prev then
          refreshed = refreshed + 1
          -- Refresh upstream-authoritative fields only. Upstream can RESOLVE a thread,
          -- but a local resolve is never reverted by re-import (would lose user intent).
          prev.body = nullable(cm.body) or prev.body
          prev.file, prev.side = path, side
          prev.line_start, prev.line_end = line, line
          prev.author = nullable(cm.author) and nullable(cm.author.login) or prev.author
          prev.github_thread_id, prev.origin = t.id, "github"
          prev.reactions = reaction_counts(nullable(cm.reactionGroups) or {})
          if t.isResolved or prev.status == "resolved" then prev.status = "resolved"
          elseif t.isOutdated then prev.status = "outdated"
          else prev.status = "draft" end
          local rev = side == "LEFT" and source:base_rev() or source:head_rev()
          local lines = anchor.file_lines(meta.repo_root, rev, path) or {}
          prev.anchor = anchor.compute(lines, math.min(line, math.max(1, #lines)))
          if side == "LEFT" then prev.anchor.blob_sha = rev end
          prev.updated_at = now
          if i == 1 then root_local_id = prev.id end
          store.comments[prev.id] = prev
        else
          local rev = side == "LEFT" and source:base_rev() or source:head_rev()
          local lines = anchor.file_lines(meta.repo_root, rev, path) or {}
          local a = anchor.compute(lines, math.min(line, math.max(1, #lines)))
          if side == "LEFT" then a.blob_sha = rev end
          local comment = {
            id = util.uuid(), source_key = store.source_key, file = path, side = side,
            head_sha = source:head_rev(), base_sha = source:base_rev(),
            line_start = line, line_end = line, anchor = a, rename_lineage = {},
            kind = (nullable(cm.body) or ""):find("```suggestion") and "suggestion" or "normal",
            body = nullable(cm.body) or "", origin = "github",
            status = t.isResolved and "resolved" or (t.isOutdated and "outdated" or "draft"),
            github_id = gid, github_thread_id = t.id,
            in_reply_to = (i > 1) and root_local_id or nil,
            author = nullable(cm.author) and (nullable(cm.author.login) or "github") or "github",
            reactions = reaction_counts(nullable(cm.reactionGroups) or {}),
            created_at = nullable(cm.createdAt) or now, updated_at = now, hidden = false,
          }
          store.comments[comment.id] = comment
          if i == 1 then root_local_id = comment.id end
          imported = imported + 1
        end
      end
    end
  end
  store:save()
  return imported, nil, refreshed
end

return M
